src/Guzzle/MiddlewareFactory.php

```diff
@@ -81,38 +93,38 @@ private function newRetryDecider() {
 			}
 
 			if( $response ) {
-				$headers = $response->getHeaders();
 				$data = json_decode( $response->getBody(), true );
 
 				// Retry on server errors
 				if( $response->getStatusCode() >= 500 ) {
 					$shouldRetry = true;
 				}
 
-				if ( array_key_exists( 'mediawiki-api-error', $headers ) ) {
-					foreach( $headers['mediawiki-api-error'] as $mediawikiApiErrorHeader ) {
-						if (
-							// Retry if we have a response with an API error worth retrying
-							in_array(
-								$mediawikiApiErrorHeader,
-								array(
-									'ratelimited',
-									'readonly',
-									'internal_api_error_DBQueryError',
-								)
+				foreach( $response->getHeader( 'Mediawiki-Api-Error' ) as $mediawikiApiErrorHeader ) {
+					if (
+						$response->getHeaderLine( 'Retry-After' )
+						||
+						// Retry if we have a response with an API error worth retrying
+						in_array(
+							$mediawikiApiErrorHeader,
+							array(
+								'ratelimited',
+								'maxlag',
+								'readonly',
+								'internal_api_error_DBQueryError',
 							)
-							||
-							// Or if we have been stopped from saving as an 'anti-abuse measure'
-							// Note: this tries to match "actionthrottledtext" i18n messagae for mediawiki
-							(
-								$mediawikiApiErrorHeader == 'failed-save' &&
-								strstr( $data['error']['info'], 'anti-abuse measure' )
-							)
-						) {
-							$shouldRetry = true;
-						}
-
+						)
+						||
+						// Or if we have been stopped from saving as an 'anti-abuse measure'
+						// Note: this tries to match "actionthrottledtext" i18n messagae for mediawiki
+						(
+							$mediawikiApiErrorHeader == 'failed-save' &&
+							strstr( $data['error']['info'], 'anti-abuse measure' )
+						)
+					) {
+						$shouldRetry = true;
```

Comment:
This method is really long... though that was the case before already and you did not make it worse